### DISCUSSION TITLE
Automated Extraction of POS, TAG, and Dependency tags?

### DISCUSSION THREAD

**Discussion by awindsor** [explosion/spaCy#11623] [Help: Model Advice] [answered]
In order to accommodate the switching of models and changing tag sets I would like to be able to extract three sets of tags from the model

1. The coarse grained POS tags
2. The fine grained TAG tags
3. The dependency tags. 

I realize that 2 and 3 are available from the model homepage (although 1 is not) but I would like to adapt automatically to changing models and not have to manually enter them from the homepage.

**Comment by polm** [comment, marked as answer]
You can generally get the labels of a component like this, using the tagger as an example:

```
labels = nlp.get_pipe("tagger").labels
```

This should work for the tagger and parser, for example. 

The coarse grained tags are [Universal Depencies tags](https://universaldependencies.org/u/pos/), plus SPACE, and do not change based on pipeline. 

If you have a component that doesn't have a list of labels, like a Morphologizer or an AttributeRuler, then you can iterate over a bunch of data to pretty reliably get the list of labels. Depending on the specific component there may also be other ways to recover the labels.

Can you clarify how you're using the labels?

**Reply by awindsor** [reply]
Thank you. 

labels = nlp.get_pipe("tagger").labels
labels = nlp.get_pipe("parser").labels

is exactly what I wanted. I am using this in two ways:

As I change language models I have code which flags every line that uses a dependency or tagger tag that is not present in the new language model. In some cases the fix is simply a matter of adding tags to a global constant and the intersecting that with the list of currently available tags. In other cases, the issues are linguistically more complicated and requires real work.
